stactask/task.py

```diff
@@ -64,12 +65,13 @@ def __init__(
         skip_upload: bool = False,
         skip_validation: bool = False,
     ):
+        # define these to avoid undefined in destructor if init fails
+        self._save_workdir = True
+        self._workdir = Path()
+
```

Comment:
With `self._workdir = Path()` and noting that `str(Path()) == "."` and `bool(Path()) == True`, it seems like if `getLogger` or `self.validate` fail, then `cleanup_workdir` will attempt to `rmtree(".")`, regardless of what value was provided in `workdir`.

Failing a better idea, it seems like `self._workdir` should default to `None` so that the guard condition in `cleanup_workdir` will avoid `rmtree(".")`.

Reply:
I agree with @ircwaves, but I think the necessity of setting these values here is actually an indication of a different problem.

Rather than having cleanup performed in two places now (destructor _and_ handler), I think the cleanup should only be performed in _one_ place, the handler. Doing so indicates that the creation of the workdir should also move exclusively to the handler. With such a change, the destructor doesn't have to care about cleanup: temp workdirs are scoped to the duration of the handler execution and not beyond.

More concretely, these lines should be something like (feel free to use better names):

```suggestion
        # define these to avoid undefined in destructor if init fails
        self._save_workdir_setting = save_workdir
        self._workdir_setting = workdir

```

Then, lines 84-93 move into a new `maybe_create_workdir` method, which is called at the beginning of the `handler` method, with updates for the new instance property names `self._save_workdir_setting` and `self._workdir_setting` as the input values from the user.

Alternatively, `save_workdir` and `workdir` could become `handler` args, given that their use is scoped to just the handler, but having just those wouldn't be backward compatible. I think it might be good to support both the instance-level and `handler`-level specification of these parameters, perhaps even going so far as to deprecate the instance-level ones.

Reply:
Another, perhaps cleaner, way to express the creation/cleanup would be with a new workdir class that encapsulates the logic/operations around directory creation and saving, then make that a `contexthandler`-enabled class so it can be used in the `handler` via a `with` statement, rather than the `try`/`finally`.

Reply:
Thanks for the comments on this. I decided to do the following:

1) replace `__del__` with requiring that cleanup_workdir be called
2) handler is going to be the majority use case for this (e.g., not many people are going to be creating a Task directly), so handler is responsible for calling cleanup_workdir. 
3) `__init__` still handles creating of the workdir, as this is **necessary** to do work in `process`, but cleaning up is an optional action after it's done

Reply:
Oh, right, `handler` is a `classmethod`. That's what I get for trying to quickly review this between other things, I miss stuff like that. And that is exactly why I abandoned working on this change awhile back, because the complexity of the "right" answer quickly spirals.

I still think that the problems here point to the wrong abstractions. Is it actually that `Task` should be the context handler with cleanup of the directory happening within the exit method? Perhaps that means `Task` is subclass of the `workdir` class I proposed. Then, in the handler, you do `with cls(payload, **kwargs) as task:` instead of the `try`/`finally`.

All this said, I think these changes are the first step to enabling the context handler idea, and the revisions do remove the issue pointed out by @ircwaves. So ✅ from me.

Reply:
Yeah, I agree -- there should be some contextual abstraction around the workdir handling, and Task shouldn't be responsible for managing it, only knowing where it is.

Reply:
That does the trick.  And the complication is awkward....but doing people favors (creating work directories for them) never goes unpunished.